src/io.c

```diff
@@ -196,10 +241,60 @@ Obj FuncIO_RestoreSIGCHLDHandler( Obj self )
   }
 }
 
+// The following function checks if a PID is marked as ignored.
+// Returns 1 if the PID was ignored, 0 otherwise.
+static int IO_CheckForIgnoredPid( int pid )
+{
+    int i;
+    // Make sure a new signal doesn't come in while looking at array
+    signal(SIGCHLD, SIG_DFL);
+    int found = 0;
+    for(i = 0; i < ignoredpidslen; ++i) {
+        if (ignoredpids[i] == pid) {
+            ignoredpids[i] = ignoredpids[ignoredpidslen - 1];
+            ignoredpidslen--;
+            found = 1;
+            break;
+        }
+    }
+    signal(SIGCHLD,IO_SIGCHLDHandler);
```

Comment:
That looks wrong: this function is called by `IO_HandleChildSignal` which in turn is called by `FuncIO_WaitPid` which also disables the signal handler. With the above code, you restore the signal handler to `IO_SIGCHLDHandler` prematurely.

Similarly, it is called by the signal handler `IO_SIGCHLDHandler` itself, with the same problem.

So I think the two signal calls must be removed here; and I'd add another comment at the top of the function, warning that it must only be called if the signal handler is disabled.